### PR TITLE
cljc version of math.combinatorics

### PR DESCRIPTION
**Pull request by jurgenphotek**
this is not actually ready yet. but would you be interested?

This would be 1.7+ only.

### CODE REVIEW

**Comment by Engelberg** [comment]
1. Contrib projects can't accept pull requests, but require a contributor agreement on file, and a patch filed through the JIRA system.

2. This is a question that a lot of the contrib projects are wrestling with, because cljc is nice to have, but kills backwards compatibility.  I'd like to see a couple other contrib projects go first in terms of making the switch, but in principle, I'd like to see this happen.  I think this should probably port to cljc rather trivially.

**Comment by puredanger** [comment]
Please note that currently Clojure contrib projects with cljc cannot be built in the Clojure CI system (due to a stack of yaks). The only project currently using cljc is test.check, which must be built through a painful manual process. Because of that, I would prefer to delay this until that issue is fixed.

We have mapped out the path to fixing this issue but it's been delayed due to other work. It is considered an important issue to fix and we do intend to do it.

**Comment by thedavidmeister** [comment]
this sounds like a good idea :)

**Comment by viebel** [comment]
@puredanger Is it possible now to have Clojure contrib projects with `cljc`?

File: src/main/clojure/clojure/math/combinatorics.cljc [hunk 1]
```diff
@@ -9,194 +9,188 @@
      :doc "Efficient, functional algorithms for generating lazy
 sequences for common combinatorial functions. (See the source code 
 for a longer description.)"}
-  clojure.math.combinatorics
-  (:refer-clojure :exclude [update]))
+  clojure.math.combinatorics
+  (:refer-clojure :exclude [update])
+  #?(:cljs (:require [goog.string :refer [format]])))
```
Comment (viebel) [inline src/main/clojure/clojure/math/combinatorics.cljc]: @jurgenphotek the correct way of using `goog.string.format` is by requiring `goog.string.format` like this: `(:require goog.string.format)`

**Comment by viebel** [comment]
Now, we can have contrib projects with`cljc` https://groups.google.com/forum/#!topic/clojure-dev/cwxzA3K-P3k

@jurgenphotek @Engelberg do you want to update the PR?

**Comment by seancorfield** [comment]
@viebel Note the earlier comment here that Contrib projects cannot accept PRs -- changes must go through JIRA (and be covered by a signed contributor's agreement).

**Comment by jurgenphotek** [comment]
Looks like @Engelberg has this under way: http://dev.clojure.org/jira/browse/MCOMB-6

**Comment by Engelberg** [comment]
I have some time to take care of this over the next couple of days.  Is
there anything I need to know about the way the new build system handles
cross-platform code, or should it be as simple as getting a cljc file
working and then running the build and deploy cycle via the same web tool
as before?

**Comment by puredanger** [comment]
The first step should be that "mvn clean package" builds the correct archive on your local machine.

I believe I have made the necessary changes to the build system and the project config such that it should "just work" to then deploy on the build box as before. But when that is found to be tragically wrong, please let me know. :)

**Comment by Engelberg** [comment]
This is the current version of pom.xml in the project:
https://github.com/clojure/math.combinatorics/blob/master/pom.xml

It appears to have been updated 23 days ago (related to cross-platform
support?), but I don't see anything in the file related to cross-platform
support.
https://github.com/clojure/math.combinatorics/blob/master/pom.xml

**Comment by Engelberg** [comment]
I've managed to get something working by cobbling together pieces from the
pom.xml in clojure.data.xml.  I now have something that is building and
passing tests in both Clojure and Clojurescript on my machine.  My pom.xml
now looks like this (question below, so please read past the pom.xml to see
the question):

```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="
http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="
http://maven.apache.org/POM/4.0.0
http://maven.apache.org/xsd/maven-4.0.0.xsd">
  <modelVersion>4.0.0</modelVersion>
  <artifactId>math.combinatorics</artifactId>
  <version>0.1.4-SNAPSHOT</version>
  <name>math.combinatorics</name>

  <parent>
    <groupId>org.clojure</groupId>
    <artifactId>pom.contrib</artifactId>
    <version>0.2.0</version>
  </parent>

  <developers>
    <developer>
      <name>Mark Engelberg</name>
    </developer>
  </developers>

  <dependencies>
    <dependency>
      <groupId>org.clojure</groupId>
      <artifactId>clojurescript</artifactId>
      <version>1.9.293</version>
      <scope>test</scope>
    </dependency>
  </dependencies>

  <build>
    <testResources>
      <testResource>
        <directory>${project.basedir}/src/test/clojurescript</directory>
      </testResource>
    </testResources>
    <plugins>
      <plugin>
        <!-- cljc-maven-plugin produces .clj files from .cljc files,
             hence enabling usage of .cljc while supporting clojure < 1.7.0
-->
        <groupId>net.bendlas</groupId>
        <artifactId>cljc-maven-plugin</artifactId>
        <version>0.1.2</version>
        <executions>
          <execution>
            <id>split-compile</id>
            <phase>generate-sources</phase>
            <goals><goal>split</goal></goals>
          </execution>
          <execution>
            <id>split-compile-tests</id>
            <phase>generate-test-sources</phase>
            <goals><goal>split-tests</goal></goals>
          </execution>
        </executions>
      </plugin>
      <plugin>
        <artifactId>maven-jar-plugin</artifactId>
        <version>2.4</version>
        <executions>
          <execution>
            <id>default-jar</id>
            <phase>package</phase>
            <goals>
              <goal>jar</goal>
            </goals>
            <configuration>
              <includes>
                <include>**/*.clj</include>
                <include>**/*.cljs</include>
                <include>**/*.cljc</include>
              </includes>
            </configuration>
          </execution>
        </executions>
      </plugin>
    </plugins>
  </build>

  <properties>
    <clojure.version>1.4.0</clojure.version>
  </properties>

  <pluginRepositories>
    <!-- use clojars for cljc-maven-plugin -->
    <pluginRepository>
      <id>clojars.org</id>
      <url>https://clojars.org/repo</url>
    </pluginRepository>
  </pluginRepositories>

  <scm>
    <connection>scm:git:git@github.com:
clojure/math.combinatorics.git</connection>
    <developerConnection>scm:git:git@github.com:
clojure/math.combinatorics.git</developerConnection>
    <url>git@github.com:clojure/math.combinatorics.git</url>
    <tag>HEAD</tag>
  </scm>
</project>
```

I'd like to specifically ask about:
```
  <properties>
    <clojure.version>1.4.0</clojure.version>
  </properties>
```

I am assuming that this designates the minimum version of Clojure that this
works with.  Is that correct?  While making changes so it would be
cross-platform code, I intentionally made some changes that I knew would
break support for 1.2.  I have no idea why it no longer works with 1.3
(complains about not being able to find `deftest` which doesn't make much
sense to me because I think deftest has been around much longer than
that).  So I've marked the minimum version as 1.4.0.

Is it safe to assume that when I push it and run the build tool that it
will pick up on that property and only test from 1.4 up in the build-matrix?

**Comment by bendlas** [comment]
The `clojure.version` property is the one, that the package gets compiled with, i.e. the one that will generate .class files. Lower versions won't be able to load it because of clojure's ABI changes, but the test matrix will have to be adjusted separately.

**Comment by Engelberg** [comment]
On Fri, Jan 6, 2017 at 1:42 AM, Herwig Hochleitner <notifications@github.com
> wrote:

> The clojure.version property is the one, that the package gets compiled
> with, i.e. the one that will generate .class files. Lower versions won't be
> able to load it because of clojure's ABI changes, but the test matrix will
> have to be adjusted separately.
>
>
So I might as well put 1.8.0 in the clojure.version property then.  But how
do I make that adjustment to the test matrix?

**Comment by bendlas** [comment]
> So I might as well put 1.8.0 in the clojure.version property then.  But how
> do I make that adjustment to the test matrix?

As I said, lower versions of clojure won't be able to load class files generated by higher versions (the ABI is about classes and methods on a java level), so putting 1.8.0 will mean that your minimum requirement is 1.8.0

As for the test matrix, @puredanger helped adjust that for data.xml. Depending on your access level on jenkins, you might be able to do it yourself.

**Comment by puredanger** [comment]
FYI, you should not need cljc-maven-plugin to do a basic cljc build and I'd prefer not to include it if there is no reason to use it, as it's just more stuff to maintain.

I can adjust the test.matrix - everything on the build box is generated from a data file (https://github.com/clojure/build.ci/blob/master/ci_data.clj), so any manual adjustments there will get blown away.

**Comment by Engelberg** [comment]
Thanks for reclarifying, Herwig.  I think I understand your point now.  I
misunderstood and thought you were advocating 1.8 to make sure I was
getting the highest-quality and speediest class files, but you were saying
it wouldn't be backwards compatible if I did that.

So the main reason to use the cljc-maven-plugin (aside from the fact that I
wasn't getting a working build prior to including it) is that it allows the
library to be backwards compatible to 1.4.  I'm all for supporting back
versions if it is easy to do so.

But Alex, if you'd prefer for me to just use the existing pom.xml and only
support 1.7 and up, I'll do it, but I'll need a little more instruction
about what to do next with the pom.xml since the build triggered by that
one seems to still be looking for the clj file and not finding it.

**Comment by puredanger** [comment]
@Engelberg I think it's ok to make Clojure 1.7 the min requirement and not use cljc-maven-plugin and that would be my preference. Users for <1.7 can still use the prior version if needed.

Can you commit stuff to a branch or put a patch somewhere so I can look at it?

**Comment by puredanger** [comment]
I just pushed changes to master that convert to cljc. For me, this builds and tests locally and produces a correct jar file. Someone should actually test it on ClojureScript though.

**Comment by puredanger** [comment]
This also builds and tests properly on the build box. Someone needs to test the clojurescript version before including it in a release though.

**Comment by bendlas** [comment]
> advocating 1.8 to make sure I was
> getting the highest-quality and speediest class files, but you were saying
> it wouldn't be backwards compatible if I did that.

Hm, when looking at the jar file distributed through maven, there are no `.class` files. That means, that the failure I observed when upping `clojure.version` either was due to a (subsequently fixed) incompatibility in the project, or caused by the test-matrix using an AOT version (which could be fixed there). Either way `clojure.version` should just determine what is run by `mvn test`. Sorry for the confusion.

`cljc-maven-plugin` is meant as a stop-gap and proof-of-concept. I created a [ticket](https://github.com/talios/clojure-maven-plugin/issues/105) for `clojure-maven-plugin`.

@puredanger it seems you remain unconvinced, that pre-`1.7` + cljc compatibility is worth maintaining build tooling for. I have no qualms recognizing `1.7` as a minimum targed for "reasonably recent" clojure, based on recent survey numbers (in fact I could rewrite a couple of expressions in data.xml in a much more elegant way). If that's the message we want to send though, why even run test matrices for older versions?

**Comment by puredanger** [comment]
@bendlas by default, most of the contribs are not distributed as AOT-compiled jars. However, the default parent pom is configured to AOT compile the project as a check to ensure that it compiles (see https://github.com/clojure/build.poms/blob/master/pom.xml#L112-L114). 

The clojure.version is used when performing the main project build on the build box. The test-matrix project will test all Clojure and JDK versions that are specified as supported in the ci_data file (https://github.com/clojure/build.ci/blob/master/ci_data.clj). 

I dislike doing source pre-processing. The whole point of having cljc files is to avoid doing source pre-processing like we did with cljx. I also dislike including more plugins as that just adds more infrastructure that needs to be understood, maintained, and updated. 

My belief based on the state of the overall ecosystem is that the overwhelming majority of users are using Clojure 1.7+ and thus using cljc files is a sufficient solution. For those that are not, they can continue using older versions of the library.

**Comment by Engelberg** [comment]
I pushed my version using the cljc-maven-plugin to a branch called cljc for
you to look at.

P.S. Git does crazy, bizarre things to the line endings whenever I deal
with contrib source code.  I'm on Windows, and the recommended setting on
Windows is for git to convert endings to CRLF when checking things out and
convert back to LF when checking things in.  Despite having things
configured in this recommended way, I get bizarre diffs on contrib
projects, some of which show additional carriage returns and some of which
show that I've touched every line when I haven't.  I think it may have to
do with the patch-based system we use in contrib, which could be
circumventing the usual line-ending logic.  I *hope* that despite the weird
things I see on this end, git is committing my changes with LF line
endings.  If you see weirdness on your end when browsing the branch, and
see an easy way to clean it up, let me know.

**Comment by Engelberg** [comment]
P.P.S. I *have* tested my branch and know that it works on the latest
version of clojurescript.

**Comment by Engelberg** [comment]
The diff looks clean on the github side:
https://github.com/clojure/math.combinatorics/compare/cljc?expand=1

Maybe I need to completely blow away my local repo and clone it fresh.

**Comment by Engelberg** [comment]
Alex, I updated my pom.xml to match your modifications today (going back to
the basic one with no extra maven plugin) and it is now building on my
machine.

So, I think I'm ready to push all these changes to master and build a new
release with your simpler pom.xml.  I'll wait a little while to give anyone
who wants to look at it a chance to peruse the branch I just pushed.

**Comment by bendlas** [comment]
@puredanger alright, thanks for clarifying that. I'll conform to your stance in data.xml and remove `cljc-maven-plugin`. TBH, I'd rather invest that time into adding a [cljs test runner](https://github.com/talios/clojure-maven-plugin/issues/99) to `clojure-maven-plugin`, too.

**Comment by puredanger** [comment]
@bendlas I'd also much rather have cljs test runner support :)

**Comment by Engelberg** [comment]
Implemented in 0.1.4

**Pull request closed without merging**